src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java

```diff
@@ -0,0 +1,167 @@
+package org.broadinstitute.hellbender.utils.iterators;
+
+import htsjdk.samtools.SAMFileHeader;
+import org.broadinstitute.hellbender.GATKBaseTest;
+import org.broadinstitute.hellbender.engine.Shard;
+import org.broadinstitute.hellbender.engine.ShardBoundary;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.read.ArtificialReadUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class ShardingIteratorUnitTest extends GATKBaseTest {
```

Comment:
This iterator needs more/better tests, in addition to the tests you have below. In particular, I think you need to:

* Test with reads of different lengths, sorted by start position. As noted above, I think you have a bug in your iterator in this case.

* Test with realistically-sized (~75 bp or longer) reads that overlap, instead of just 1-base non-overlapping reads.

Reply:
Does the realistically-sized test adds something to the unit test itself? I think that the itegration test with realistic BAM files is enough.

I will definitely implement the other test case before fixing the bug to demonstrate the issue and that the fix works.